taskmap/taskmap.py

```diff
@@ -113,33 +137,47 @@ def run_parallel_async(graph, nprocs=None, sleep=0.2):
                 else:
                     cpuq.put(task)
 
+            if trap_exceptions == False:
+                is_alive = sum([p.is_alive() for p in processes]) == len(processes)
+                if not is_alive:
+                    failed = [p for p in processes if not p.is_alive()]
+                    msg = 'pid {}: failed tasks {}'.format(os.getpid(), failed)
+                    mlog(graph).error(msg)
+                    map(lambda x: x.terminate(), processes)
+                    raise Exception(msg)
             time.sleep(sleep)
 
         return tgraph.recover_values_from_manager(graph)
 
 
-def run_scheduler(graph, sleep, ioq, cpuq):
+def run_scheduler(graph, sleep, ioq, cpuq, trap_exceptions=True):
     loop = asyncio.new_event_loop()
-    loop.run_until_complete(scheduler(graph, sleep, ioq, cpuq, loop))
+    loop.run_until_complete(scheduler(graph, sleep, ioq, cpuq, loop, trap_exceptions))
     loop.close()
 
 
 # TODO: scheduler can be improved
-async def scheduler(graph, sleep, ioq, cpuq, loop):
-    while not tgraph.all_done(graph):
-        try:
-            task = ioq.get_nowait()
-            log(graph).info('pid {}: dequeueing task {}'.format(os.getpid(), task))
-            asyncio.ensure_future(run_task_async(graph, task), loop=loop)
-        except Exception:
+async def scheduler(graph, sleep, ioq, cpuq, loop, trap_exceptions):
+    if trap_exceptions:
+        while not tgraph.all_done(graph):
             try:
-                task = cpuq.get_nowait()
+                task = ioq.get_nowait()
                 log(graph).info('pid {}: dequeueing task {}'.format(os.getpid(), task))
-                asyncio.ensure_future(run_task_async(graph, task), loop=loop)
-                # don't put two cpu intensive tasks on the same core without waiting
-                await asyncio.sleep(sleep)
-            except:
-                await asyncio.sleep(sleep)
+                asyncio.ensure_future(run_task_async(graph, task, True), loop=loop)
+            except Exception:
+                try:
+                    task = cpuq.get_nowait()
+                    log(graph).info('pid {}: dequeueing task {}'.format(os.getpid(), task))
+                    asyncio.ensure_future(run_task_async(graph, task, True), loop=loop)
+                    # don't put two cpu intensive tasks on the same core without waiting
+                    await asyncio.sleep(sleep)
+                except:
+                    await asyncio.sleep(sleep)
+    else:
+        while not tgraph.all_done(graph):
+            task = ioq.get_nowait()
```

Comment:
It looks like you never pull tasks from the cpuq if `trap_exceptions=False`.  Note also that `get_nowait` throws if the queue is empty, but `get` blocks until something comes into the queue. This is why the existing code first tries to get_nowait out of th ioq, then moves on to the cpuq, then waits to try again.

You can think of the ioq queue as containing tasks that are io bound and the cpuq queue as containing tasks that are cpu bound.  The user can specify which tasks are io bound at graph creation time.  This is used so we can get the io bound tasks started first, so we don't spend X time on a cpu intensive task, start an io bound task later and then just have the cpu sitting unused while we wait for the io bound task.